### ISSUE TITLE
Waagent cannot handle gzipped CustomData prep'd by Terraform

### ISSUE THREAD

**Issue by u6006** [Azure/WALinuxAgent#1559]
One of our deployment failed with the following error in waagent.log and the reason is that the CustomData passed in is generated by Terraform and it's gzipped before base64 encoded. Both our version (v2.2.25) and the latest have no capability to handle it since it tries to decode UTF-8 after base64 decoding.

Please ignore it if there's the same issue previously reported.

2019/06/17 11:09:40.519282 INFO Decode custom data
...
  File "/usr/lib/python2.7/site-packages/azurelinuxagent/common/osutil/default.py", line 1047, in decode_customdata
    return base64.b64decode(data).decode('utf-8')
  File "/usr/lib/python2.7/encodings/utf_8.py", line 16, in decode
    return codecs.utf_8_decode(input, errors, True)
UnicodeDecodeError: 'utf8' codec can't decode byte 0x8b in position 1: invalid start byte
, duration=0

**Comment by trstringer** [comment]
You're correct, waagent will only decode the data and doesn't handle gzipped custom data. Looking at the Terraform documentation for [azurerm_virtual_machine](https://www.terraform.io/docs/providers/azurerm/r/virtual_machine.html) there is no mentioned of Terraform gzipping the custom_data first.

I also see [where Terraform base64 encodes the custom data](https://github.com/terraform-providers/terraform-provider-azurerm/blob/33174e76022dba277d58eadbf02430ce57a8daf2/azurerm/resource_arm_virtual_machine.go#L1497-L1500), but I don't see the code path that compresses it? Perhaps this might be a good issue to open up with the Terraform provider.
